server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -99,12 +101,24 @@ public class CoordinationDiagnosticsService implements ClusterStateListener {
     // Non-private for testing
     volatile ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> clusterFormationResponses = null;
 
+    /*
+     * This is a reference to the task that is periodically reaching out to a master eligible node to get its CoordinationDiagnosticsResult
+     * for diagnosis. It is null when no polling is occurring.
+     * The field is accessed (reads/writes) from multiple threads, and is also reassigned on multiple threads.
+     */
+    volatile AtomicReference<Scheduler.Cancellable> remoteStableMasterHealthIndicatorTask = null;
+    /*
+     * This field holds the result of the task in the remoteStableMasterHealthIndicatorTask field above. The field is accessed
+     * (reads/writes) from multiple threads, and is also reassigned on multiple threads.
+     */
+    volatile AtomicReference<RemoteMasterHealthResult> remoteCoordinationDiagnosisResult = null;
```

Comment:
Does this need to be an `AtomicReference` or a `volatile RemoteMasterHealthResult` would suffice?

Reply:
For the same reason as the task AtomicReference, we really need this to be a reference. If I change it to just a RemoteMasterHealthResult, then each polling stack no longer gets its own copy so it is possible that a polling stack that we want to have been cancelled overwrites the result value of the current polling stack.

Reply:
Ah gotcha - we'd need an empty instance of `RemoteMasterHealthResult` to simulate the same thing. Thanks for the explanation !